code/modules/mob/living/silicon/ai/ai_say.dm

```diff
@@ -49,108 +49,167 @@
+				var/changed_name = lowertext(replacetext(character_name, " ", "_"))
+				if(allowed_characters[changed_name])
+					passed_characters += character_name
+		else
+			passed_characters = usable_characters
+	else
+		if(banned_characters.len)
+			passed_characters = usable_characters
+			for(var/character_name in usable_characters)
+				var/changed_name = lowertext(replacetext(character_name, " ", "_"))
+				if(banned_characters[changed_name])
+					passed_characters -= character_name
+		else
+			passed_characters = usable_characters
+	var/character_to_use = input(src, "Choose what 15.ai character to use:", "15.ai Character Choice")  as null|anything in passed_characters
+	if(!character_to_use)
+		return
+	var/max_characters = 300 // magic number but its the cap 15 allows
+	var/message = input(src, "Use the power of 15.ai to say anything! ([max_characters] character maximum)", "15.ai VOX System", src.last_announcement) as text|null
 
 	if(!message || announcing_vox > world.time)
 		return
 
-	last_announcement = message
-
 	if(incapacitated())
 		return
 
 	if(control_disabled)
 		to_chat(src, "<span class='warning'>Wireless interface disabled, unable to interact with announcement PA.</span>")
 		return
 
-	var/list/words = splittext(trim(message), " ")
-	var/list/incorrect_words = list()
-
-	if(words.len > 30)
-		words.len = 30
-
-	for(var/word in words)
-		word = lowertext(trim(word))
-		if(!word)
-			words -= word
-			continue
-		if(!GLOB.vox_sounds[word])
-			incorrect_words += word
-
-	if(incorrect_words.len)
-		to_chat(src, "<span class='notice'>These words are not available on the announcement system: [english_list(incorrect_words)].</span>")
+	if(length(message) > max_characters)
+		to_chat(src, "<span class='notice'>You have too many characters! You used [length(message)] characters, you need to lower this to [max_characters] or lower.</span>")
 		return
+	var/regex/check_for_bad_chars = regex("\[^a-zA-Z!?.,' :\]+")
+	if(check_for_bad_chars.Find(message))
+		to_chat(src, "<span class='notice'>These characters are not available on the 15.ai system: [english_list(check_for_bad_chars.group)].</span>")
+		return
+	last_announcement = message
 
 	announcing_vox = world.time + VOX_DELAY
 
-	log_game("[key_name(src)] made a vocal announcement with the following message: [message].")
-	log_talk(message, LOG_SAY, tag="VOX Announcement")
-
-	for(var/word in words)
-		play_vox_word(word, src.z, null)
-
-
-/proc/play_vox_word(word, z_level, mob/only_listener)
-
-	word = lowertext(word)
-
-	if(GLOB.vox_sounds[word])
-
-		var/sound_file = GLOB.vox_sounds[word]
-		var/sound/voice = sound(sound_file, wait = 1, channel = CHANNEL_VOX)
-		voice.status = SOUND_STREAM
-
-	// If there is no single listener, broadcast to everyone in the same z level
+	log_game("[key_name(src)] started making a 15.AI announcement with the following message: [message]")
+	message_admins("[key_name(src)] started making a 15.AI announcement with the following message: [message]")
```

Comment:
could this include which voice they're using in the logging, and (I don't know if this is do-able) but maybe a cancel button?